### ISSUE TITLE
File Descriptor sometimes leak with consult

### ISSUE THREAD

**Issue by kamahen** [SWI-Prolog/swipl-devel#622]
V8.3.2 (V8.3.1 is OK)

Confirmed using both strace(1) and Prolog trace/2 on open/{3,4}, close{1,2}. I'm trying to reproduce with a smaller test case, and also am instrumenting code to narrow down the conditions that leak FDs.

**Comment by JanWielemaker** [comment]
This issue has been mentioned on **SWI-Prolog**. There might be relevant details there:

https://swi-prolog.discourse.group/t/ann-swi-prolog-8-3-2/2536/4

**Comment by kamahen** [comment]
The first commit where the FD leak occurs:
git difftool --tool=tkdiff cf00d089bcef49e5a00fd410f4dda0613601e27e 528d0b03e5fca71a42f8438164d9cb2c0d80634c

**Comment by kamahen** [comment]
I've looked at some of the code, and it seems to have a matching close for every open, guaranteed by a `setup_call_cleanup`. (Adding instrumentation to `setup_call_cleanup` seems problematic, so I'm putting that off for now.) There are a few bits of code that might be questionable (such as a call to `erase/1` before the close in` $close_source`; the `erase/1` call could silently fail) but so far, tracing doesn't show a problem with those (e.g., all calls to `erase/1` seem to succeed).

I tried turning off multi-threading (including gc); this made the code run incredibly slowly, but still showed the "Too many open files" error.

**Comment by kamahen** [comment]
The way to trigger this appears to be: dynamically execute a `load_files` predicate, and the loaded file contains `:-set_prolog_flag(autoload,false).`. When that directive was commented out, everything worked fine.

When autoload is turned off, it appears that `library(error)` is autoloaded (!) and that gets into a loop of further loads (including `library(error)`). I'm looking further into this.

**Comment by kamahen** [comment]
The bug appears to be triggered by a `qlf` program dynamically loading a file that contains a directive to turn off autoload.

Here's a simple way to reproduce the problem:

* Down load the files [z4.pl.txt](https://github.com/SWI-Prolog/swipl-devel/files/4822368/z4.pl.txt) and [z3.pl.txt](https://github.com/SWI-Prolog/swipl-devel/files/4822369/z3.pl.txt) and rename to z4.pl and z3.pl.
* Compile: `swipl --stand_alone=true --undefined=error --verbose=true --foreign=save -o z4.qlf -c z4.pl`
* To see what should be output: `swipl z4.pl`
* To show the crash: `./z4.qlf`
It will spew out many lines of "compiled" messages and eventually crash:
```
[Thread 1 (main) at Tue Jun 23 16:10:40 2020] ../src/pl-srcfile.c:1501: endReconsult: Assertion failed: reload->pred_access_count == popNPredicateAccess(0)
C-stack trace labeled "assert_fail":
  [0] save_backtrace() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/os/pl-cstack.c:332 [0x7faf7162dd88]
  [1] __assert_fail() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/pl-assert.c:103 [0x7faf715eaa47]
  [2] endReconsult() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/pl-srcfile.c:1501 (discriminator 1) [0x7faf715da08c]
  [3] pl_fixup_reconsult1_va() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/pl-srcfile.c:1656 [0x7faf715dad1b]
  [4] PL_next_solution() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/pl-vmi.c:3810 [0x7faf7155304a]
  [5] PL_call_predicate() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/pl-fli.c:4146 [0x7faf715626ed]
  [6] propagateAutoload() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/os/pl-prologflag.c:576 [0x7faf7162faeb]
  [7] pl_set_prolog_flag2_va() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/os/pl-prologflag.c:981 [0x7faf71630a92]
  [8] PL_next_solution() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/pl-vmi.c:3810 [0x7faf7155304a]
  [9] query_loop() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/pl-pro.c:144 [0x7faf7159fb74]
  [10] prologToplevel() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/pl-pro.c:485 [0x7faf715a041b]
  [11] PL_initialise() at /home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/../src/pl-init.c:1107 [0x7faf715e1f13]
  [12] ./z4.qlf(+0x766) [0x563d7d86c766]
  [13] __libc_start_main() at /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:344 [0x7faf71142b97]
  [14] ./z4.qlf(+0x7ba) [0x563d7d86c7ba]
Aborted (core dumped)
```

**Comment by kamahen** [comment]
I also ran the test under strace; for the first attempt to read library(error), it didn't try error.qlf but only error.pl ... for subsequent (looping) calls, it did try error.qlf first.

BTW, in this sequence stat() is called 3 times (1 call should suffice, and only if there's a check for .qlf vs .pl); and it's not clear to me why there's any need to look at the directory or to do an access() call.
```
stat("/home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/home/library", {st_mode=S_IFDIR|0755, st_size=12288, ...}) = 0
stat("/home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/home/library/error.pl", {st_mode=S_IFREG|0644, st_size=14990, ...}) = 0
stat("/home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/home/library/error.pl", {st_mode=S_IFREG|0644, st_size=14990, ...}) = 0
access("/home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/home/library/error.pl", R_OK) = 0
stat("/home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/home/library", {st_mode=S_IFDIR|0755, st_size=12288, ...}) = 0
stat("/home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/home/library/error.pl", {st_mode=S_IFREG|0644, st_size=14990, ...}) = 0
openat(AT_FDCWD, "/home/peter/src/swipl-devel/build-V8.3.2-3-gd0d08885e/home/library/error.pl", O_RDONLY) = 5
ioctl(5, TCGETS, 0x7fff844d2180)        = -1 ENOTTY (Inappropriate ioctl for device)
fcntl(5, F_SETFD, FD_CLOEXEC)           = 0
read(4, "/*  Part of SWI-Prolog\n\n    Auth"..., 4096) = 4096
```

**Comment by kamahen** [comment]
Also, shouldn't --autoload when compiling prevent any lookup in /usr/lib/swi-prolog/library? (e.g., error.pl shows in the verbose output from compilation, so there should be no need to look outside the z4.qlf file)

**Comment by kamahen** [comment]
Fixed by 21e0a70e91d7aa9c5dd83b00f0d49e22c1ed8b60

If I can reproduce the other issues I mentioned, I'll open new isseus.
